caddyhttp/proxy/reverseproxy.go

```diff
@@ -260,8 +260,23 @@ func (rp *ReverseProxy) ServeHTTP(rw http.ResponseWriter, outreq *http.Request,
 		}
 		defer backendConn.Close()
 
-		go pooledIoCopy(backendConn, conn) // write tcp stream to backend
-		pooledIoCopy(conn, backendConn)    // read tcp stream from backend
+		// Proxy backend -> frontend.
+		go pooledIoCopy(conn, backendConn)
+
+		// Proxy frontend -> backend.
+		//
+		// NOTE: Hijack() sometimes returns buffered up bytes in brw which
+		// would be lost if we didn't read them out manually below.
+		if brw != nil {
+			if n := brw.Reader.Buffered(); n > 0 {
+				rbuf, err := brw.Reader.Peek(n)
+				if err != nil {
+					return err
+				}
+				backendConn.Write(rbuf)
```

Comment:
This implementation looks correct.

It could be simplified by avoiding the check if `n > 0` since its okay to `Peek` zero bytes:
```go
rbuf, err := brw.Reader.Peek(brw.Reader.Buffered())
```